### ISSUE TITLE
support worker threads

### ISSUE THREAD

**Issue by coranos** [vrza/node-blake2#36]
I'm trying to call blake2 from a worker thread, and im getting the error:

`Error: Module did not self-register: 'node_modules/blake2/build/Release/binding.node'.`

it looks like the fix is simple based on this SO post:

https://stackoverflow.com/questions/68598248/is-there-a-way-to-run-native-code-modules-using-worker-threads

replace NODE_MODULE with NODE_MODULE_WORKER_ENABLED

so I tried to implement the fix but get an error building it

https://github.com/vrza/node-blake2/compare/master...coranos:node-blake2:master

```
../src/binding.cpp:259:1: error: C++ requires a type specifier for all declarations
NODE_MODULE_WORKER_ENABLED(binding, init)
^
../src/binding.cpp:259:28: error: use of undeclared identifier 'binding'
NODE_MODULE_WORKER_ENABLED(binding, init)
                           ^
../src/binding.cpp:259:42: error: expected ';' after top level declarator
NODE_MODULE_WORKER_ENABLED(binding, init)
```

I'm new to node-gyp so I apologize if I'm making a noob mistake trying to get worker threads working.

**Comment by vrza** [comment]
@coranos according to that SO post, it should be

NAN_MODULE_WORKER_ENABLED

rather than

NODE_MODULE_WORKER_ENABLED

With this change it still builds (at least on my Linux x86_64 machine) and passes all unit tests.

Please do test it for your use case and report back, but off the bat I have no objections to making this change.

**Comment by coranos** [comment]
it works, so i created a PR.

https://github.com/vrza/node-blake2/pull/37

one wierd thing though, if I increase MAX_WORKERS from 2 to 10 it segfaults:

https://github.com/vrza/node-blake2/pull/37/files#diff-254d406659798a45ac16ff52bce5544e88f98c9e97a8eb0d8616caac9f138a4fR7

sh: line 1:  4916 Segmentation fault: 11  node node_modules/mocha/bin/mocha tests

I'm not sure what would cause it, but it does limit how useful it is in my use case.

**Comment by vrza** [comment]
Sure enough, I can reproduce the segfault, the stack trace is below. It seems to segfault when a  worker threads tries to load the addon, during addon registration. From initial investigation I can only surmise that this addon (the node binding, not the underlying BLAKE2 implementation) is not thread-safe.

```
#0  std::__shared_ptr<v8::internal::Counters, (__gnu_cxx::_Lock_policy)2>::get (this=0xd589) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/shared_ptr_base.h:1666
#1  0x00005555587160bd in v8::internal::Isolate::counters (this=0x1) at ../deps/v8/src/execution/isolate.h:1102
#2  0x0000555558c9e588 in v8::internal::GlobalHandles::NodeSpace<v8::internal::GlobalHandles::Node>::Free (this=0x7ffd28278a40, node=0x7ffd282aa2f0) at ../deps/v8/src/handles/global-handles.cc:299
#3  v8::internal::GlobalHandles::NodeSpace<v8::internal::GlobalHandles::Node>::Release (node=0x7ffd282aa2f0) at ../deps/v8/src/handles/global-handles.cc:288
#4  0x0000555558c9e3c8 in v8::internal::GlobalHandles::Destroy (location=0x7ffd282aa2f0) at ../deps/v8/src/handles/global-handles.cc:1132
#5  0x00005555587006e5 in v8::api_internal::DisposeGlobal (location=0x7ffd282aa2f0) at ../deps/v8/src/api/api.cc:887
#6  0x00007ffff7fae176 in node_register_module_v108 () from /home/random/github/coranos/node-blake2/build/Release/binding.node
#7  0x00005555581b3a50 in node::binding::DLOpen(v8::FunctionCallbackInfo<v8::Value> const&)::$_0::operator()(node::binding::DLib*) const (this=0x7ffd44347ab0, dlib=0x7ffd44344f10) at ../src/node_binding.cc:485
#8  0x00005555581b3790 in std::__invoke_impl<bool, node::binding::DLOpen(v8::FunctionCallbackInfo<v8::Value> const&)::$_0&, node::binding::DLib*>(std::__invoke_other, node::binding::DLOpen(v8::FunctionCallbackInfo<v8::Value> const&
)::$_0&, node::binding::DLib*&&) (__f=..., __args=@0x7ffd56ff2ee0: 0x7ffd44344f10) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/invoke.h:61
#9  0x00005555581b373d in std::__invoke_r<bool, node::binding::DLOpen(v8::FunctionCallbackInfo<v8::Value> const&)::$_0&, node::binding::DLib*>(node::binding::DLOpen(v8::FunctionCallbackInfo<v8::Value> const&)::$_0&, node::binding::
DLib*&&) (__fn=..., __args=@0x7ffd56ff2ee0: 0x7ffd44344f10) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/invoke.h:114
#10 0x00005555581b3635 in std::_Function_handler<bool (node::binding::DLib*), node::binding::DLOpen(v8::FunctionCallbackInfo<v8::Value> const&)::$_0>::_M_invoke(std::_Any_data const&, node::binding::DLib*&&) (__functor=..., 
    __args=@0x7ffd56ff2ee0: 0x7ffd44344f10) at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/std_function.h:290
#11 0x000055555811a32a in std::function<bool (node::binding::DLib*)>::operator()(node::binding::DLib*) const (this=0x7ffd56ff3278, __args=0x7ffd44344f10)
    at /usr/lib/gcc/x86_64-pc-linux-gnu/12/include/g++-v12/bits/std_function.h:591
#12 0x00005555580ed0b4 in node::Environment::TryLoadAddon(char const*, int, std::function<bool (node::binding::DLib*)> const&) (this=0x7ffd442a88f0, 
    filename=0x7ffd56ff3388 "/home/random/github/coranos/node-blake2/build/Release/binding.node", flags=1, was_loaded=...) at ../src/env.cc:531
#13 0x00005555581b17c4 in node::binding::DLOpen (args=...) at ../src/node_binding.cc:450
#14 0x0000555558865e0b in v8::internal::FunctionCallbackArguments::Call (this=0x7ffd56ff4180, handler=...) at ../deps/v8/src/api/api-arguments-inl.h:147
#15 0x00005555588644d9 in v8::internal::(anonymous namespace)::HandleApiCallHelper<false> (isolate=0x7ffd44000e60, function=..., new_target=..., fun_data=..., receiver=..., args=...) at ../deps/v8/src/builtins/builtins-api.cc:112
#16 0x00005555588608a5 in v8::internal::Builtin_Impl_HandleApiCall (args=..., isolate=0x7ffd44000e60) at ../deps/v8/src/builtins/builtins-api.cc:142
#17 0x000055555885febb in v8::internal::Builtin_HandleApiCall (args_length=7, args_object=0x7ffd56ff4850, isolate=0x7ffd44000e60) at ../deps/v8/src/builtins/builtins-api.cc:130
#18 0x000055555a37a179 in Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit ()
#19 0x000055555a2fe0d0 in Builtins_InterpreterEntryTrampoline ()
```

Solving for the worker threads use case could be somewhat involved. For reference, node.js has some [documentation](https://nodejs.org/api/addons.html#addons_context_aware_addons) on the topic of thread-safe add-ons.

**Comment by coranos** [comment]
ok so good news bad news.

good news is I found what was causing the code to not be threadsafe. The copy function uses a static variable. Bad news is I am unsure how to rewrite the copy to be thread safe, as I'm not familiar with what it is trying to do.

i dont know if you have any insight as to why a static variable is required. ive never seen someoen copy a hash context before.

i updated my code so it can run with 100 worker threads, but copy doesnt work, and the tests are commented out.

I'll try to think more about Copy. thank you for your insight so far :)

**Comment by vrza** [comment]
That's more good news than bad news, we are making progress! I was able to replace that global constructor variable with a local variable inside methods that use it (Init and Copy). All tests are passing!

In order to release this we will need to drop support for older versions of node that don't support worker threads. This change also warrants bumping the major version of this addon (to 5.0).

Created a new feature branch to wrap up the release and CI stuff:

https://github.com/vrza/node-blake2/tree/worker-threads

**Comment by vrza** [comment]
All done, created new pull request #38 

Thank you @coranos! In the new version I would like to add you to the contributors section of package.json, please let me know your full name and/or email address, or if you'd prefer to be mentioned just by your github account handle, or even if you'd like to not be mentioned.

**Comment by coranos** [comment]
Ah just mention my account handle. thank you very much.

**Comment by vrza** [comment]
Released [v5.0.0](https://github.com/vrza/node-blake2/releases/tag/v5.0.0) with this functionality
